libusb/os/windows_winusb.c

```diff
@@ -1480,6 +1480,110 @@ static int set_hid_interface(struct libusb_context *ctx, struct libusb_device *d
 	return LIBUSB_SUCCESS;
 }
 
+static int get_guid(struct libusb_context *ctx, char *dev_id, HDEVINFO *dev_info, SP_DEVINFO_DATA *dev_info_data,
+					int guid_number, GUID **if_guid)
+{
+	DWORD size, reg_type;
+	HKEY key;
+	char *guid_string, *new_guid_string;
+	char *guid;
+	LONG s;
+	size_t len;
+	int i;
+	int err = LIBUSB_SUCCESS;
+
+	if (!pSetupDiGetDeviceRegistryPropertyA(*dev_info, dev_info_data, SPDRP_DRIVER,
+		NULL, NULL, 0, NULL) && (GetLastError() != ERROR_INSUFFICIENT_BUFFER)) {
+		usbi_info(ctx, "The following device has no driver: '%s'", dev_id);
+		usbi_info(ctx, "libusb will not be able to access it");
+	}
+	// ...and to add the additional device interface GUIDs
+	key = pSetupDiOpenDevRegKey(*dev_info, dev_info_data, DICS_FLAG_GLOBAL, 0, DIREG_DEV, KEY_READ);
+	if (key == INVALID_HANDLE_VALUE) {
+		usbi_warn(ctx, "Cannot get the additional GUIDs for '%s'", dev_id);
+		return LIBUSB_ERROR_ACCESS;
+	}
+
+	size = MAX_GUID_STRING_LENGTH + 1;
+	guid_string = malloc(size);
+	if (guid_string == NULL) {
+		usbi_err(ctx, "failed to alloc guid_string");
+		return LIBUSB_ERROR_NO_MEM;
+	}
+	for(i = 0; i < 2; i++) {
+		*guid_string = '\0';
+		// Look for both DeviceInterfaceGUIDs *and* DeviceInterfaceGUID, in that order
+		// If multiple GUIDs just process the first and ignore the others
+		s = pRegQueryValueExA(key, "DeviceInterfaceGUIDs", NULL, &reg_type,
+			(LPBYTE)guid_string, &size);
+		if (s == ERROR_FILE_NOT_FOUND)
+			s = pRegQueryValueExA(key, "DeviceInterfaceGUID", NULL, &reg_type,
+				(LPBYTE)guid_string, &size);
+		if ( s == ERROR_SUCCESS) {
+			//the GUID was read successfully
+			break;
+		}
+		else if (s == ERROR_FILE_NOT_FOUND) {
+			usbi_warn(ctx, "no DeviceInterfaceGUID registered for '%s'", dev_id);
+			err = LIBUSB_ERROR_ACCESS;
+			goto exit;
+		} else if (s == ERROR_MORE_DATA) {
+			new_guid_string = realloc((void *)guid_string, size);
+			if (new_guid_string == NULL) {
+				usbi_err(ctx, "failed to realloc guid string");
+				err = LIBUSB_ERROR_NO_MEM;
+				goto exit;
+			}
+			// Try to read GUID again
+			guid_string = new_guid_string;
+		} else {
+			usbi_warn(ctx, "unexpected error from pRegQueryValueExA for '%s'", dev_id);
+			err = LIBUSB_ERROR_ACCESS;
+			goto exit;
+		}
+	}
+
+	// https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regqueryvalueexa#remarks
+	// - "string may not have been stored with the proper terminating null characters"
+	// - "Note that REG_MULTI_SZ strings could have two terminating null characters"
+	if (reg_type == REG_SZ && size > 1 && guid_string[size - 1] == 0) {
+		guid = guid_string;
+	} else if (reg_type == REG_MULTI_SZ && size > 2 && guid_string[size - 1] == 0 && guid_string[size - 2] == 0) {
+		/* get the n-th guid indexed by guid_number since the DeviceInterfaceGUIDs may
```

Comment:
... and this doesn't allow for only one terminating null character any longer, right?